contracts/LimitOrderSwap.sol

```diff
@@ -70,9 +70,64 @@ contract LimitOrderSwap is ILimitOrderSwap, Ownable, TokenCollector, EIP712 {
         _fillLimitOrder(order, makerSignature, takerParams, true);
     }
 
+    /// @inheritdoc ILimitOrderSwap
+    function fillLimitOrderGroup(
+        LimitOrder[] calldata orders,
+        bytes[] calldata makerSignatures,
+        uint256[] calldata makerTokenAmounts,
+        address[] calldata profitTokens,
+        uint256 unwrapAmount
+    ) external payable override {
+        if (orders.length != makerSignatures.length || orders.length != makerTokenAmounts.length) revert InvalidParams();
+
+        // validate orders and calculate takingAmounts
+        uint256[] memory takerTokenAmounts = new uint256[](orders.length);
+        for (uint256 i = 0; i < orders.length; ++i) {
+            LimitOrder calldata order = orders[i];
+            uint256 makingAmount = makerTokenAmounts[i];
+
+            (bytes32 orderHash, uint256 orderFilledAmount) = _validateOrder(order, makerSignatures[i]);
+            {
+                uint256 orderAvailableAmount = order.makerTokenAmount - orderFilledAmount;
+                if (makingAmount > orderAvailableAmount) revert NotEnoughForFill();
+                takerTokenAmounts[i] = ((makingAmount * order.takerTokenAmount) / order.makerTokenAmount);
+
+                if (makingAmount == 0 && takerTokenAmounts[i] == 0) revert ZeroTokenAmount();
+
+                // record fill amount
+                orderHashToMakerTokenFilledAmount[orderHash] = orderFilledAmount + makingAmount;
+            }
+
+            // collect maker tokens
+            _collect(order.makerToken, order.maker, address(this), makingAmount, order.makerTokenPermit);
+
+            // transfer fee if present
+            uint256 fee = (makingAmount * order.feeFactor) / Constant.BPS_MAX;
+            order.makerToken.transferTo(feeCollector, fee);
+
+            _emitLimitOrderFilled(order, orderHash, takerTokenAmounts[i], makingAmount - fee, fee, address(this));
+        }
+
+        // unwrap WETH if trader specified
+        if (unwrapAmount > 0) {
+            weth.withdraw(unwrapAmount);
+        }
```

Comment:
nit: I feel that a flag is more convenient for sender, since he just needs to decide to receive `ETH` or `WETH`, and doesn't need to know exactly amount, and this block also introduce a new possibility to revert for incorrect amount.

Though when sender has the ability to find the ring trade, I think he also has the ability to count the `ETH` and `WETH` profits beforehand. But I'm not sure if there is really a case that the sender must to receive with partial `ETH` and partial `WETH` in one group fill.

Reply:
Will have a discussion.